🎉New Source: Jina AI for web scraping and feeding LLM models

btkcodedev opened the pull request:
## What
Closes https://github.com/airbytehq/PyAirbyte-Hackathon/issues/32
* New Source: Jina AI
* Made with Low Code
* Builder Compatible
* Parameters macro not used
* Poetry flow

## How
* New Source builder with referring: https://jina.ai/reader/#apiform

## Review guide
1. `manifest.yaml`
2. `docs/sources/jina-ai.md`

## User Impact
* Enhancement

## Can this PR be safely reverted and rolled back?
- [x] YES 💚

btkcodedev commented:
All tests are passing the local
Please ping me on Slack for config JSON
![image](https://github.com/airbytehq/airbyte/assets/47392334/0784a3ab-f9fb-4d51-84a6-eca5e383d99a)

btkcodedev commented:
Please ping @btkcodedev in slack for credentials

aaronsteers commented on airbyte-integrations/connectors/source-jina-ai/README.md:
```diff
@@ -0,0 +1,103 @@
+# Jina Ai Source
```
```suggestion
# Jina AI Source
```

aaronsteers commented on airbyte-integrations/connectors/source-jina-ai/source_jina_ai/manifest.yaml:
```diff
@@ -0,0 +1,158 @@
+      - api_key
+    additionalProperties: true
+    properties:
+      api_key:
+        type: string
+        title: API Key
+        description: >-
+          Your Jina AI API Key. See <a
+          href="https://jina.ai/reader/#apiform">here</a>.
+        airbyte_secret: true
```
Is it possible to make this optional? I think we get a lower rate limit for the requests, but it can streamline POCs and smaller use cases if so.

btkcodedev replied:
:white_check_mark:

aaronsteers commented on airbyte-integrations/connectors/source-jina-ai/source_jina_ai/manifest.yaml:
```diff
@@ -0,0 +1,158 @@
+          href="https://jina.ai/reader/#apiform">here</a>.
+        airbyte_secret: true
+        order: 0
+      read_prompt:
+        type: string
+        title: Your read prompt url, that would be appended at the end of reader stream
+        default: "https://www.google.com"
+        order: 1
```
Can we accept an array of URLs strings?

And if so, perhaps a rename to match the use case? Maybe something like `reader_urls` or `reader_page_urls`?

If too difficult to implement, let me know. I understand it will change the endpoint and "get" protocol to make this change.

I imagine that the same `reader` stream could produce multiple records in each sync, with the URL serving as primary key for each record.

btkcodedev replied:
Array of URL's is not supported I guess, the [platform](https://jina.ai/reader/#apiform) example is also taking one at a time.
:thinking:

aaronsteers replied:
@btkcodedev - I did some digging and I may have found a workaround for this issue. What do you think of this approach?

1. We would use the "parameterized requests" option in the No-Code UI Builder.
2. We'd pass an array input from the user config. (Example in screenshot below.)
3. We'd use the value from the parameterized input for the request itself, instead of passing directly from the config.

![image](https://github.com/airbytehq/PyAirbyte-Hackathon/assets/18150651/fd2f7528-794f-4b4f-8d6b-48896e17aee0)

Totally okay if this doesn't work (I didn't have a chance to test it), but if you are able to give it a shot and report back, that's much appreciated. 🙏

btkcodedev replied:
@aaronsteers I've tried it but has a catch, the parametrized request is the ListPartitionRouter in lowcode
the injection of that partition can be only done inside body, parameters or headers but not on path or url

Ref: 
![image](https://github.com/airbytehq/airbyte/assets/47392334/d4335495-c77d-4b3a-a4d5-0191e4664875)

So failed on that one

Error Ref: `Jinja macro has undeclared variables: {'stream_partition'}.`
![image](https://github.com/airbytehq/airbyte/assets/47392334/279cea68-96e7-45ca-b0b7-cc49cb821bd6)

btkcodedev replied:
Sorryy :bow:

aaronsteers commented on airbyte-integrations/connectors/source-jina-ai/source_jina_ai/manifest.yaml:
```diff
@@ -0,0 +1,158 @@
+        title: Your read prompt url, that would be appended at the end of reader stream
+        default: "https://www.google.com"
+        order: 1
+      search_prompt:
+        type: string
+        title: >-
+          Your search prompt, that would be appended at the end of search stream, Should be an encoded string without spaces.
+          Visit https://coding.tools/url-encode for seamless encoding
+        default: "Search%20airbyte"
+        order: 2
```
I'm less familiar with this endpoint, but I was thinking similarly as above: perhaps we could make plural and accept `search_prompts` as an array of search strings?

Also - from a usability perspective (maybe also for security/stability), what do you think of requesting an unencoded string like "who is the current president?" and resolving (internally) to something like "who%20is%20the%20current%20president"?

btkcodedev replied:
Same as URL, it can't accept multiple search prompts I guess
we could use the connector with multiple configs in a loop with pyairbyte for making the affect :)

aaronsteers commented on airbyte-integrations/connectors/source-jina-ai/source_jina_ai/manifest.yaml:
```diff
@@ -0,0 +1,158 @@
+          Visit https://coding.tools/url-encode for seamless encoding
+        default: "Search%20airbyte"
+        order: 2
+      gather_links:
+        type: boolean
+        title: >-
+          Set this as true for creating "Buttons & Links" section at the end. This helps the downstream LLMs or web agents navigating 
+          the page or take further actions.
+        order: 3
+      gather_images:
+        type: boolean
+        title: >-
+          Set this as true for creating "Images" section at the end. This gives the downstream LLMs an overview of all visuals on the page, 
```
Nice. I like that these options exist - great for customization.

btkcodedev replied:
:bow:

aaronsteers commented on docs/integrations/sources/jina-ai.md:
```diff
@@ -0,0 +1,51 @@
+
+## Overview
+
+Jina AI Reader API addresses these issues of scraping by extracting the core content from a URL and converting it into clean, LLM-friendly text, ensuring high-quality input for your agent and RAG systems.
+
+### Available Streams
```
Suggesting a quick explanation of both APIs for framing at the top of this readme. Something like this?:

```suggestion

This connector allows access to the [Jina Reader API](https://jina.ai/reader/) using two modes:

- **"Reader" Mode** (`https://r.jina.api`) - Given a set of one or more URLs as input, return the content of those pages as Markdown text. The Reader endpoint extracts the core content from a URL and converting it into clean, LLM-friendly text, ensuring high-quality input for your agent and RAG systems.
- **"Search" Mode** (`https://s.jina.api`) - Similar to the reader endpoint, but accepting a search prompt and returning the text from top 5 search results.

Both of these API endpoints will generate human readable markdown, which can also be efficiently processed by downstream LLM and GenAI applications.

Both modes can be utilized in the same sync, following the configuration instructions below.

### Available Streams
```

btkcodedev replied:
:white_check_mark:

aaronsteers commented on docs/integrations/sources/jina-ai.md:
```diff
@@ -0,0 +1,51 @@
+# Jina AI
```
I think the title should probably have "Reader" in the name since (I believe) both the Search and Reader endpoints are advertised under "Reader" brand. (Also because Jina has several other AI services we are not trying to encapsulate here.)

```suggestion
# Jina AI Reader
```

btkcodedev replied:
:white_check_mark:

aaronsteers commented on airbyte-integrations/connectors/source-jina-ai/metadata.yaml:
```diff
@@ -0,0 +1,40 @@
+data:
+  allowedHosts:
+    hosts:
+      - "https://*.jina.ai"
```
I wonder if we should just list these two endpoints explicitly.

```suggestion
    hosts:
      - "https://s.jina.ai"
      - "https://r.jina.ai"
```

btkcodedev replied:
:white_check_mark:

aaronsteers reviewed (changes requested):
@btkcodedev - This is looking really great! I left comments inline - feel free to push back if you have counter-proposals or if you are thinking of it differently. 🙏

btkcodedev commented on airbyte-integrations/connectors/source-jina-ai/README.md:
```diff
@@ -0,0 +1,103 @@
+# Jina Ai Source
+
+This is the repository for the Jina Ai configuration based source connector.
```
:white_check_mark:

btkcodedev commented on airbyte-integrations/connectors/source-jina-ai/README.md:
```diff
@@ -0,0 +1,103 @@
+# Jina Ai Source
```
:white_check_mark:

btkcodedev commented:
@aaronsteers All comments resolved,
- Made bearer token optional - Added custom component
- Added search prompt auto encoder - Added config migration
- Added unit tests for config migration

All works

btkcodedev commented:
CI tests (Local): All pass
Ref:
![image](https://github.com/airbytehq/airbyte/assets/47392334/eb9e89ed-05bd-4c5d-b91e-29aa82ba1534)


Also, GSM creds have reached it's limit, Please update the credentials to these (Remove api_key) 
{
  "read_prompt": "https://www.google.com",
  "search_prompt": "What is AI",
  "gather_links": true,
  "gather_images": true
}

aaronsteers commented on airbyte-integrations/connectors/source-jina-ai-reader/metadata.yaml:
```diff
@@ -0,0 +1,42 @@
+    - suite: unitTests
+    - suite: acceptanceTests
+      testSecrets:
+        - name: SECRET_SOURCE-JINA-AI-READER__CREDS
+          fileName: config.json
+          secretStore:
+            type: GSM
+            alias: airbyte-connector-testing-secret-store
```
Note to self. We'll want to register a secret here for Airbyte-side integration tests.

cc @aaronsteers

aaronsteers replied:
Logged here: https://github.com/airbytehq/airbyte-internal-issues/issues/8399

aaronsteers reviewed (approved):
Thanks for these changes and for attempting the array approach. I'm circling your findings back internally in the hope that a future version of the Builder may support arrays as inputs here.

🚀 🙌
